### PR TITLE
[sdk upgrade] 

### PR DESCRIPTION
**Pull request by mayankamencherla**
- [x] Minor fixes to the readme file
- [x] Clean up code
- [x] Test Cases
- [x] Test each new feature , and the existing features
- [ ] Mock out the test responses, esp Tokens
- [ ] Code Review

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -23,9 +22,8 @@ else
 Usage
 -----
 ### Initialize
-`
-RazorpayClient client = new RazorpayClient(key, secret);
-`
+`RazorpayClient client = new RazorpayClient(key, secret);`
```
Comment (harman28) [inline README.md]: Doing this means no syntax highlighting. I suggest we keep it as it is, and throw in `csharp` as the language identifier.

File: README.md [hunk 2]
```diff
@@ -13,8 +13,7 @@ If you are using nuget package manager, you can add below in your packages.confi
 
 `<package id="Razorpay" version="1.0.0" targetFramework="net40" />`  
 or  
-`<package id="Razporpay" version="1.0.0" targetFramework="net45" />
-`
+`<package id="Razporpay" version="1.0.0" targetFramework="net45" />`
```
Comment (harman28) [inline README.md]: And `xml` here.

Comment (sunnyagg) [inline README.md]: can you fix the version number to latest version

File: src/RazorpayClient.cs [hunk 3]
```diff
@@ -1,18 +1,21 @@
...
+
 namespace Razorpay.Api
 {
     public class RazorpayClient
     {
-        public const string Version = "1.0.0";
-        public const string BaseUrl = "https://api.razorpay.com/v1/";
-        public static string Key = null;
-        public static string Secret = null;
+        protected const string Version = "1.0.0";
```
Comment (sunnyagg) [inline src/RazorpayClient.cs]: update version to 1.2.0

File: src/RazorpayClient.cs [hunk 4]
```diff
@@ -1,18 +1,57 @@
...
+        private Invoice invoice = null;
 
         public RazorpayClient(string key, string secret)
         {
             RazorpayClient.Key = key;
-            RazorpayClient.Secret = secret;
+            RazorpayClient.Secret = secret;
+        }
+
+        public static string getKey()
```
Comment (sunnyagg) [inline src/RazorpayClient.cs]: Read about Properties in C#, its much cleaner way to expose the attributes

File: src/RazorpayClient.cs [hunk 5]
```diff
@@ -1,18 +1,57 @@
...
+        public static void setAppsDetails(string title, string version)
+        {
+            Dictionary<string, string> appDetail = new Dictionary<string, string>();
+            appDetail.Add("title", title);
+            appDetail.Add("version", version);
+
+            appsDetails.Add(appDetail);
+        }
+
+        public static List<Dictionary<string, string>> getAppsDetails()
```
Comment (sunnyagg) [inline src/RazorpayClient.cs]: same here as well

File: README.md [hunk 6]
```diff
@@ -11,10 +11,15 @@ Installation
...
-`<package id="Razorpay" version="1.0.0" targetFramework="net40" />`  
+```xml
+<package id="Razorpay" version="1.2.0" targetFramework="net40" />
+```  
+
 or  
-`<package id="Razporpay" version="1.0.0" targetFramework="net45" />
-`
+
+```cs
```
Comment (amitmahbubani) [inline README.md]: this is `xml`?

File: src/Entity.cs [hunk 7]
```diff
@@ -100,6 +100,30 @@ private Entity ParseEntity(dynamic response)
...
+            }
+            else if (response["entity"] == "card")
+            {
+                entity = new Card();
+            }
+            else if (response["entity"] == "transfer")
+            {
+                entity = new Transfer();
+            }
+            else if (response.Property("transfer_id") != null)
```
Comment (amitmahbubani) [inline src/Entity.cs]: why expect `transfer_id`?

File: src/Payment.cs [hunk 8]
```diff
@@ -43,5 +43,27 @@ public Refund Refunds
...
+        }
+
+        public List<Transfer> Transfers(Dictionary<string, object> options = null)
+        {
+            string relativeUrl = string.Format("payments/{0}/{1}", this["id"], GetEntityUrl());
+            List<Entity> entities = Request(relativeUrl, HttpMethod.Get, null);
+
+            List<Transfer> transfers = new List<Transfer>();
+
+            foreach(Entity entitiy in entities)
```
Comment (amitmahbubani) [inline src/Payment.cs]: typo `entity`

File: src/Payment.cs [hunk 9]
```diff
@@ -43,5 +43,27 @@ public Refund Refunds
...
+        public Transfer Transfer(Dictionary<string, object> data = null)
+        {
+            string relativeUrl = GetEntityUrl() + "/" + this["id"] + "/transfers";
+            List<Entity> entities = Request(relativeUrl, HttpMethod.Post, data);
+            return (Transfer)entities[0];
+        }
+
+        public List<Transfer> Transfers(Dictionary<string, object> options = null)
+        {
+            string relativeUrl = string.Format("payments/{0}/{1}", this["id"], GetEntityUrl());
```
Comment (amitmahbubani) [inline src/Payment.cs]: `relativeUrl` becomes `payments/:id/payments` ?

File: src/RazorpayClient.cs [hunk 10]
```diff
@@ -1,18 +1,82 @@
...
+        protected const string baseUrl = "https://api.razorpay.com/v1/";
+        protected static string key = null;
+        protected static string secret = null; 
+        protected static List<Dictionary<string, string>> appsDetails = new List<Dictionary<string, string>>();
         private Payment payment = null;
         private Order order = null;
+        private Customer customer = null;
+        private Invoice invoice = null;
+        private Token token = null;
+        private Card card = null;
```
Comment (amitmahbubani) [inline src/RazorpayClient.cs]: Define `Transfer` and `Reversal` too

File: test/RazorpayClientTest45.csproj [hunk 11]
```diff
@@ -33,8 +33,20 @@
...
     <Compile Include="Program.cs" />
     <Compile Include="Helper.cs" />
+    <Compile Include="ClientTestCases.cs" />
     <Compile Include="OrderTestCases.cs" />
     <Compile Include="PaymentTestCases.cs" />
+    <Compile Include="UtilsTestCases.cs" />
+    <Compile Include="CustomerTestCases.cs" />
+    <Compile Include="InvoiceTestCases.cs" />
+    <Compile Include="TokenTestCases.cs" />
+    <Compile Include="CardTestCases.cs" />
```
Comment (amitmahbubani) [inline test/RazorpayClientTest45.csproj]: Transfer, Reversals tests please

File: README.md (hunk 6 above)
Comment (sunnyagg) [inline README.md]: yes, most .net config files are xmls

File: src/Entity.cs [hunk 12]
```diff
@@ -100,6 +100,30 @@ private Entity ParseEntity(dynamic response)
...
             }
+            else if (response["entity"] == "customer")
+            {
+                entity = new Customer();
+            }
+            else if (response["entity"] == "invoice")
+            {
+                entity = new Invoice();
+            }
+            else if ((response["entity"] == "token") || (response.Property("deleted") != null))
```
Comment (sunnyagg) [inline src/Entity.cs]: why is this deleted check needed ?

File: src/Entity.cs [hunk 13]
```diff
@@ -100,6 +100,30 @@ private Entity ParseEntity(dynamic response)
             {
                 entity = new Order();
             }
+            else if (response["entity"] == "customer")
```
Comment (sunnyagg) [inline src/Entity.cs]: can you refactor this function instead of long if else chain ?

File: src/Refund.cs [hunk 14]
```diff
@@ -6,14 +6,14 @@ public class Refund : Entity
     {
         public string PaymentId;
 
-        public Refund Fetch(string id)
+        new public Refund Fetch(string id)
```
Comment (sunnyagg) [inline src/Refund.cs]: now we can directly fetch a refund using id, we don't need payment_id, how did we supported that in other sdks. Add that here as well

File: src/RestClient.cs [hunk 15]
```diff
@@ -2,36 +2,77 @@
...
+            return request;
+        }
+
+        private static string getAppDetailsUa()
+        {
+            List<Dictionary<string, string>> appsDetails = RazorpayClient.AppsDetails;
+
+            string appsDetailsUa = string.Empty;
+
+            foreach(Dictionary<string, string> appsDetail in appsDetails)
```
Comment (sunnyagg) [inline src/RestClient.cs]: this does not look correct, discuss

File: src/Entity.cs (hunk 7 above)
Comment (mayankamencherla) [inline src/Entity.cs]: @amitmahbubani The other api's don't have a field called `transfer_id`, and reversal response doesn't contain any entity. 

https://docs.razorpay.com/v1/page/marketplace#transfersidreversals

File: src/Payment.cs (hunk 9 above)
Comment (mayankamencherla) [inline src/Payment.cs]: @amitmahbubani Mistake. Fixed.

File: src/RazorpayClient.cs (hunk 10 above)
Comment (mayankamencherla) [inline src/RazorpayClient.cs]: @amitmahbubani But you get a transfer from the payment entity, and a reversal from the transfer entity. So why define it here as well?

File: src/Entity.cs (hunk 12 above)
Comment (mayankamencherla) [inline src/Entity.cs]: @sunnyagg When a token is deleted, the json response is simply `deleted: true`.

File: src/RazorpayClient.cs (hunk 10 above)
Comment (amitmahbubani) [inline src/RazorpayClient.cs]: Transfers and Reversals can be independent of payments too.

Comment (mayankamencherla) [inline src/RazorpayClient.cs]: @amitmahbubani But currently the only way to create a reversal is through a transfer. So, ATM, is Reversal needed in RazorpayClient.cs?

Comment (amitmahbubani) [inline src/RazorpayClient.cs]: @mayankamencherla Nope, reversals will come up later. Let's add then.

File: src/Entity.cs (hunk 7 above)
Comment (amitmahbubani) [inline src/Entity.cs]: Docs fixed.

File: README.md (hunk 6 above)
Comment (amitmahbubani) [inline README.md]: @mayankamencherla

File: src/Customer.cs [hunk 16]
```diff
@@ -0,0 +1,33 @@
...
+        }
+           
+        public Customer Edit(Dictionary<string, object> data) 
+        {
+            string relativeUrl = string.Format("{0}/{1}", GetEntityUrl(), this["id"]);
+            List<Entity> entities = Request(relativeUrl, HttpMethod.Put, data);
+            return (Customer)entities[0];
+        }
+
+        public Token Tokens()
```
Comment (sunnyagg) [inline src/Customer.cs]: this requires customer fetch, remove that condition

File: src/Entity.cs [hunk 17]
```diff
@@ -88,17 +104,16 @@ private List<Entity> Build(dynamic response)
...
+            
+            string responseEntity = (string) response["entity"];
+
+            if (Entities.ContainsKey(responseEntity) == true)
             {
-                entity = new Refund();
+                entity = Entities[responseEntity];
             }
-            else if (response["entity"] == "order")
+            else if (response.Property("deleted") != null)
```
Comment (sunnyagg) [inline src/Entity.cs]: this is not needed, just return without calling parseEntity

File: src/Payment.cs [hunk 18]
```diff
@@ -43,5 +43,30 @@ public Refund Refunds
                 return refund;
             }
         }
+
+        public Transfer Transfer(Dictionary<string, object> data = null)
```
Comment (sunnyagg) [inline src/Payment.cs]: add constructor which takes payment id and can call Transfer method, fix whereever applies

File: src/Customer.cs (hunk 16 above)
Comment (mayankamencherla) [inline src/Customer.cs]: @sunnyagg Added a constructor

File: src/Entity.cs (hunk 17 above)
Comment (mayankamencherla) [inline src/Entity.cs]: @sunnyagg Request has to return a list of entities. Therefore, I think it would be better if we can return a deleted token for this api call. Perhaps a good idea would be to manually set entity.Attributes.entity = "token" for this particular case. What do you think?

File: src/RazorpayClient.cs [hunk 19]
```diff
@@ -1,18 +1,84 @@
...
+
 namespace Razorpay.Api
 {
     public class RazorpayClient
     {
-        public const string Version = "1.0.0";
-        public const string BaseUrl = "https://api.razorpay.com/v1/";
-        public static string Key = null;
-        public static string Secret = null;
+        protected const string version = "1.2.0";
```
Comment (sunnyagg) [inline src/RazorpayClient.cs]: if breaking changes, should be 2.0

File: src/Token.cs [hunk 20]
```diff
@@ -0,0 +1,43 @@
+using System.Collections.Generic;
+
+namespace Razorpay.Api
+{
+    public class Token : Entity
+    {
+        public string CustomerId;
```
Comment (sunnyagg) [inline src/Token.cs]: it should be a constructor in customer entity!

Comment (mayankamencherla) [inline src/Token.cs]: @sunnyagg Let's discuss this once?

File: src/RazorpayClient.cs (hunk 19 above)
Comment (mayankamencherla) [inline src/RazorpayClient.cs]: @sunnyagg Not really breaking changes anymore

File: src/Entity.cs (hunk 17 above)
Comment (mayankamencherla) [inline src/Entity.cs]: @sunnyagg Lets discuss this once.

File: src/Entity.cs [hunk 21]
```diff
@@ -32,21 +48,27 @@ protected List<Entity> All(Dictionary<string, object> options = null)
...
             {
                 postData = JsonConvert.SerializeObject(options);
             }
 
             string responseStr = client.MakeRequest(relativeUrl, verb, postData);
 
+            // Check this once
+            if (verb == HttpMethod.Delete)
+            {
+                return new List<Entity>();
```
Comment (sunnyagg) [inline src/Entity.cs]: return null

File: src/Entity.cs (hunk 17 above)
Comment (sunnyagg) [inline src/Entity.cs]: @mayankamencherla this is still not removed

File: src/Transfer.cs [hunk 22]
```diff
@@ -0,0 +1,56 @@
+using System.Collections.Generic;
+
+namespace Razorpay.Api
+{
+    public class Transfer : Entity
+    {
+        public Transfer(string transferId = "")
+        {
+            this.Attributes = new Dictionary<string, object>();
```
Comment (sunnyagg) [inline src/Transfer.cs]: Move this to base as discussed.
